text/0000-portals.md

```diff
@@ -0,0 +1,108 @@
+
+Having a custom element be inserted out of its natural dom tree hierarchy would enable one to have greater control over the stacking context, allowing a predictalbe popup and modal implementation (including faux-modals that are scoped to tabs and are not per-se global).
+
+## Basic example
+
+The proposal suggests two new APIs:
+1) A way to define where in the DOM the element would be inserted (eg. `lwc:portal-definition` attribute on an element)
+2) A way to define the destination portal of an element (eg. `lwc:portal` attribute on an element)
+
+One non-trivial example of such usage:
+
+```xhtml
+<body>
+  <div lwc:portal-definition="global-modals"></div>
+
+  <div lwc:portal-definition="popups"></div>
+  ...
+  <div id="tabs">
+    <div id="tab-1">
+      <div lwc:portal-definition="tab1-modals"></div>
+
+      <div>
+         <x-dialog lwc:portal="global-modals">Global Modal</x-foo>
+         <x-dialog lwc:portal="tab1-modals">Scoped Modal</x-foo>
+      </div>  
+    </div>  
+    <div id="tab-2">
+      <div lwc:portal-definition="tab2-modals"></div>
+
+      <div>
+         <x-dialog lwc:portal="global-modals">Global Modal</x-foo>
+         <x-dialog lwc:portal="tab2-modals">Scoped Modal</x-foo>
+      </div>  
+    </div>  
+  </div>
+</body>
+```
+since we wouldn't want for a user to specify the portal each and every time one would use a dialog, more realistically -- `<x-dialog>` would have a default behaviour based on the `@Context`, which can provide whether scoping is enabled or not, the name of the scoped portal and the name of the global portal. This dynamic choice would indicate that support for dynamically specifying the portal may be needed.
+
+## Detailed design
+
+(not an actual detailed design, just adding some possible requirements)
+
+- When a portal is defined, the name must be unique.
+- Once a portal is defined, it cannot be changed (ie. `lwc:portal` can't be removed and the name is immutable). 
+- When an element that defined a portal is removed, any element inserted into it gets added back into their natural dom position. 
+- An element can't both define a portal, and define a destination portal (ie. no `lwc:portal-define` and `lwc:portal` attributes simultaneously). 
+- Instead of being inserted at its natural position in the template, the element would be inserted in the destination portal, if one exists.
+- When an element's destination portal doesn't exist, but later is inserted in the DOM, any elements that have already rendered stay in their original location, as if the portal doesn't exist.
+- When the host element is disconnected, the portalled element should also be disconnected. 
+- The portalled component should still be `@Context` aware 
+
+## Drawbacks
+
+- new API
+- additional complexity in LWC
```

Comment:
we should add the problems with the events propagation, it messes up with the mental model of the composition for authors.

Reply:
another one is the CSS issues... if the dialog is moved out, the scoped styles will not work as expected, unless we do some magic in LWC.

Reply:
i think it would be nice if events would propagate out as expected, (ie. "forwarded" to their original location), regarding CSS as discussed this potentially can be resolved if we restrict the portalled elements to be top-rooted in the template.

Reply:
I think it would be simpler and cleaner if the mental model for a portalled element is exactly the same as creating an element and appending it as a child of the portal target. The host source of the portalled element can do the following:

1) Determine the composition/tree structure of the portalled content
2) Provide the property input values for that composition (reactivity still works as expected; property changes on the host cause rehydration on the portalled content, wherever it actually lives in the document)
3) Add declarative event handlers, which are bound to methods on the host element's class, to any of the portalled elements

In this mental model, I wonder if `lwc:portal` should only apply to `template` elements. The semantics may fit better there. The contents of the `<template lwc:portal=...>...</template>` are inert (don't exist) until used. When the template is connected, the portal is located. If it exists, the contents of the template are hydrated and inserted into the portal.

Component developers assume the **contents** of a `template` with an `lwc:portal` directive are never actually in its shadow DOM, so `querySelector` from its root won't work, and event propagation won't work. This would be far more natural and unsurprising if portalled contents must be inside a `template`. That template provides the hinge for the portal itself.

Incidentally, this pattern could allow for the host component to still obtain a reference to the `template` with the `lwc:portal`. If it were important to allow the host component to query the portalled composition, it could do so from the template's reference, and LWC could provide an abstraction that fulfills that by querying the actual content where it has been inserted elsewhere in the document.